### PR TITLE
Add Protobuf support to Yokozuna

### PR DESCRIPTION
**Pull request by coderoshi**
Protocol buffers are a lower bandwidth alternative to the HTTP interface, and would be nice to implement any future clients (#37) against.

Currently this issue's aim is to match the functionality of RiakSearch. RS only supports a limited number of query params via PB, so this will mainly be of interest for existing RS users wishing to transition to Yokozuna. Also, RS only supports querying, and does not have PB support for creating/managing indexes.

### CODE REVIEW

File: riak_test/yz_languages.erl [hunk 1]
```diff
@@ -38,15 +38,6 @@ wait_for_joins(Cluster) ->
     rt:wait_until_nodes_ready(Cluster),
     rt:wait_until_no_pending_changes(Cluster).
 
-%% @doc Confirm a custom schema may be added.
```
Comment (rzezeski) [inline riak_test/yz_languages.erl:41]: I assume this test was removed since `yz_schema_admin` tests schema stuff now.  But it would have been nice to have this change in a dedicated commit for a cleaner history. Given the commit msg I have no clue why this was pulled.  And it really has nothing to do with the rest of the commit.   No need to change anything in this PR, just wanted to mention it.

File: src/yz_pb_search.erl [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    {error, missing_query};
+extract_params(#rpbsearchqueryreq{q=Query,
+                                rows=Rows, start=Start,
+                                filter=Filter,
+                                df=DefaultField, op=_DefaultOp}) ->
+    {ok, [{q, Query},
+          {wt, "json"},
+          {omitHeader, true},
+          {fq, default(Filter, "")},
+          {fl, <<"*,score">>},
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: I need to try testing this locally to verify (and we should add a test to riak-test), but we should be passing in the user supplied `fl`.  If no `fl` is supplied then we use `*,score`.  Would be good to run a few queries manually against riak search and yoko and compare results.

File: src/yz_pb_search.erl [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    Value.
+
+encode_doc(Doc) ->
+    riak_pb_search_codec:encode_search_doc([{Prop, to_binary(Value)} || {Prop, Value} <- Doc]).
+
+to_binary(A) when is_atom(A) -> to_binary(atom_to_list(A));
+to_binary(B) when is_binary(B) -> B;
+to_binary(I) when is_integer(I) -> to_binary(integer_to_list(I));
+to_binary(F) when is_float(F) -> to_binary(float_to_list(F));
+to_binary(L) when is_list(L) -> list_to_binary(L).
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: This function is suspect in regards to non ascii scripts.  It will not work for a non-ascii string.  The unicode module must be used.

```
1> S = "א בְּרֵאשִׁית, בָּרָא אֱלֹהִים, אֵת הַשָּׁמַיִם, וְאֵת הָאָרֶץ".
[1488,32,1489,1456,1468,1512,1461,1488,1513,1460,1473,1497,
 1514,44,32,1489,1464,1468,1512,1464,1488,32,1488,1457,1500,
 1465,1492,1460,1497|...]
2> 
2> B = list_to_binary(S).
** exception error: bad argument
     in function  list_to_binary/1
        called as list_to_binary([1488,32,1489,1456,1468,1512,1461,1488,1513,1460,1473,1497,
                                  1514,44,32,1489,1464,1468,1512,1464,1488,32,1488,1457,1500,
                                  1465,1492,1460|...])
3> B = unicode:characters_to_binary(S).
<<215,144,32,215,145,214,176,214,188,215,168,214,181,215,
  144,215,169,214,180,215,129,215,153,215,170,44,32,215,
  145,...>>
```

The riak test verifies that non-ascii may be stored and search, but it doesn't verify that it is retuned properly.  It would be good to add a stored field of non-ascii and verify that the same exact value is returned.

File: src/yz_app.erl [hunk 4]
```diff
@@ -24,6 +24,9 @@
 -compile(export_all).
 -include("yokozuna.hrl").
 
+%% 27 is message type rpbsearchqueryreq
+%% 28 is message type rpbsearchqueryresp
+-define(SERVICES, [{yz_pb_search, 27, 28}]).
```
Comment (rzezeski) [inline src/yz_app.erl:29]: What happens if both RS and YZ are enabled and running?  That will be the case when users upgrade.

File: riak_test/yz_pb.erl [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+-include_lib("eunit/include/eunit.hrl").
+
+-define(FMT(S, Args), lists:flatten(io_lib:format(S, Args))).
+-define(NO_HEADERS, []).
+-define(NO_BODY, <<>>).
+
+confirm() ->
+    case riak_client_exists() of
+        true ->
+            YZBenchDir = rt:get_os_env("YZ_BENCH_DIR"),
```
Comment (rzezeski) [inline riak_test/yz_pb.erl]: You're gonna want to rebase against master and change `rt:get_os_env` to `rt_config:get_os_env`.  Riak Test recently changed.  You didn't hit this problem because the verify script doesn't run a git pull.

File: src/yz_app.erl [hunk 6]
```diff
@@ -35,6 +38,7 @@ start(_StartType, _StartArgs) ->
         {ok, Pid} ->
             Routes = yz_wm_search:routes() ++ yz_wm_extract:routes() ++ yz_wm_index:routes() ++ yz_wm_schema:routes(),
             yz_misc:add_routes(Routes),
+            ok = riak_api_pb_service:register(?SERVICES),
```
Comment (rzezeski) [inline src/yz_app.erl:41]: Should this be after the service is marked up?  If a pb query lands before the service is marked up do bad things happen?

File: src/yz_pb_search.erl [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    {ok, riak_pb_codec:decode(Code, Bin)}.
+
+%% @doc encode/1 callback. Encodes an outgoing response message.
+encode(Message) ->
+    {ok, riak_pb_codec:encode(Message)}.
+
+%% @doc process/2 callback. Handles an incoming request message.
+process(Msg, #state{client=_Client}=State) ->
+    #rpbsearchqueryreq{index=IndexBin, sort=_Sort0,
+                       fl=_FL0, presort=_Presort0}=Msg,
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: No need to match against all the fields in record if they are just ignored.

File: src/yz_pb_search.erl [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    case extract_params(Msg) of
+        {ok, Params} ->
+            Mapping = yz_events:get_mapping(),
+            Index = binary_to_list(IndexBin),
+            try
+                {_Headers, Body} = yz_solr:dist_search(Index,
+                                                       [{content_type, "application/json"}],
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: The headers here are unnecessary, this is a search so no body is sent.

File: riak_test/yz_pb.erl [hunk 9]
```diff
@@ -0,0 +1,114 @@
+%% Running this test requires the `riak-erlang-client' project
```
Comment (rzezeski) [inline riak_test/yz_pb.erl]: I'm not sure if it is a recent change to riak test but I didn't need to bring in riakc.  It's under deps in riak test.

File: riak_test/yz_pb.erl [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+-module(yz_pb).
+-compile(export_all).
+-include_lib("eunit/include/eunit.hrl").
+
+-define(FMT(S, Args), lists:flatten(io_lib:format(S, Args))).
+-define(NO_HEADERS, []).
+-define(NO_BODY, <<>>).
+
+confirm() ->
+    case riak_client_exists() of
```
Comment (rzezeski) [inline riak_test/yz_pb.erl]: I say get rid of the check and bomb out hard if client isn't available.  This test should always be run.

**Comment by rzezeski** [comment]
So I was able to verify that unicode encoding does not get messed up in process of pb search.  The user just has to be aware to use `unicode:characters_to_binary` to properly convert the byte sequence.

So the `to_binary` function is probably safe BUT that's only because it is always binary type being returned for text data.

```
curl -X PUT -H 'content-type: application/json' 'http://localhost:10018/riak/encoded/stored' -d '{"name_s":"בָרָא", "text":"א בְּרֵאשִׁית, בָּרָא אֱלֹהִים, אֵת הַשָּׁמַיִם, וְאֵת הָאָרֶץ"}'

22> test_pb:go().
R: {ok,{search_results,[{"encoded",
                         [{<<"_yz_id">>,<<"test_28">>},
                          {<<"_yz_ed">>,
                           <<"20130503T033738 28 encoded test g2IHj38+">>},
                          {<<"_yz_fpn">>,<<"26">>},
                          {<<"_yz_node">>,<<"dev4@127.0.0.1">>},
                          {<<"_yz_pn">>,<<"28">>},
                          {<<"_yz_rk">>,<<"test">>},
                          {<<"_yz_rb">>,<<"encoded">>},
                          {<<"score">>,<<"1.85792129999999999779e-01">>}]},
                        {"encoded",
                         [{<<"name_s">>,
                           <<215,145,214,184,215,168,214,184,215,144>>},
                          {<<"_yz_id">>,<<"stored_49">>},
                          {<<"_yz_ed">>,
                           <<"20130503T044453 49 encoded stored g2IFGaM9">>},
                          {<<"_yz_fpn">>,<<"47">>},
                          {<<"_yz_node">>,<<"dev1@127.0.0.1">>},
                          {<<"_yz_pn">>,<<"49">>},
                          {<<"_yz_rk">>,<<"stored">>},
                          {<<"_yz_rb">>,<<"encoded">>},
                          {<<"score">>,<<"9.58915050000000018793e-02">>}]}],
                       0.18579213321208954,2}}
R2: {ok,{search_results,[],0.0,0}}
ok

23> unicode:characters_to_list(<<215,145,214,184,215,168,214,184,215,144>>).
[1489,1464,1512,1464,1488]
24> io:format("~ts~n", [unicode:characters_to_list(<<215,145,214,184,215,168,214,184,215,144>>)]).
בָרָא
```

Can't just `binary_to_list`

```
25> io:format("~ts~n", [binary_to_list(<<215,145,214,184,215,168,214,184,215,144>>)]).                        
×Ö¸×¨Ö¸×
ok
```

File: riak_test/yz_languages.erl (hunk 1 above)
Comment (coderoshi) [inline riak_test/yz_languages.erl:41]: Good point. I actually removed it because it was never called... I actually forgot I removed it.

File: src/yz_app.erl (hunk 4 above)
Comment (coderoshi) [inline src/yz_app.erl:29]: It can't be. We have the same problem with the /search WM endpoint. The client API will only allow one range of messages at a time, and to use the searchreq/resp messages, this range is it.

File: src/yz_app.erl (hunk 6 above)
Comment (coderoshi) [inline src/yz_app.erl:41]: Why would it be any different than adding wm routes before the service is up?

File: src/yz_pb_search.erl (hunk 8 above)
Comment (coderoshi) [inline src/yz_pb_search.erl]: true, this should be accept

File: src/yz_app.erl (hunk 4 above)
Comment (rzezeski) [inline src/yz_app.erl:29]: Yea, the `solr/select` end-point issue is a bug and will need to be dealt with before putting yoko in master.  This will also have to be dealt with.  But that can be a different PR, probably under #20.  Migrating users will have both systems running temporarily and we need a nice story there.

File: src/yz_pb_search.erl (hunk 2 above)
Comment (coderoshi) [inline src/yz_pb_search.erl]: I was attempting to match the current functionality of search, which doesn't actually use `fl`, though I suppose there's no danger in adding it.

File: src/yz_app.erl (hunk 6 above)
Comment (rzezeski) [inline src/yz_app.erl:41]: Ah right, the routes need to be avail before considering the yoko service up.  Same for the PB server.  I do wonder if HTTP/PB requests could come in before service is marked up and thus cause errors.  E.g. creating a coverage plan requires that the yokozuna service be marked as up, but running a HTTP/PB request requires that a coverage plan be calculated.  So a chicken and egg scenario.  I wonder if @seancribbs knows of any startup races between PB and service?

File: src/yz_pb_search.erl (hunk 8 above)
Comment (rzezeski) [inline src/yz_pb_search.erl]: Solr ignores the `accept` header.  It bases the content body solely on `wt` query value.  To add insult to injury, even when you specify `wt=json` it returns with `content-type: text/plain;charset=UTF-8`.  You can't explain that. :)

File: src/yz_pb_search.erl (hunk 2 above)
Comment (rzezeski) [inline src/yz_pb_search.erl]: Gotcha.  I think that is a slippery slope.  There are other oddities we would need to enforce:
- Can only apply `sort` when `fl` is set to unique id, otherwise must return an error (basho/riak_search#86).
- The `sort` parameter must be ignored (basho/riak_search#136).
- Possibly other weird things...

I think Yokozuna should fix these issues rather than reproduce them.  Having these issues is not a feature on which anyone should be relying on.  E.g. I don't see why someone would complain because something that used to give them an error now actually works.

File: src/yz_app.erl (hunk 6 above)
Comment (seancribbs) [inline src/yz_app.erl:41]: This problem occurs with other apps that provide PB services, so there is not really a way to address it yet. Generally, it is good enough to put your riak_api registrations after the riak_core one.

File: src/yz_pb_search.erl [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                                rows=Rows, start=Start,
+                                filter=Filter, fl=FilterList,
+                                df=DefaultField, op=DefaultOp}) ->
+    {ok, [{q, Query},
+          {wt, "json"},
+          {omitHeader, true},
+          {'q.op', default(DefaultOp, "AND")},
+          {sort, default(Sort, "")},
+          {fq, default(Filter, "")},
+          {fl, default(FilterList, <<"*,score">>)},
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: It's a field list, not filter :)

I also just realized that PB search always wants a max score.  That will only be available if `score` is passed to Solr.  So it seems we have to make sure `score` is always part of `fl`.  E.g. pretty sure a PB search with `fl=_yz_rk` would bomb out.

File: src/yz_pb_search.erl [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+            catch
+                throw:not_found ->
+                    ErrMsg = io_lib:format(?YZ_ERR_INDEX_NOT_FOUND, [Index]),
+                    {error, ErrMsg, State};
+                throw:insufficient_vnodes_available ->
+                    {error, ?YZ_ERR_NOT_ENOUGH_NODES, State};
+                throw:{Message, URL, Err} ->
+                    ?INFO("~p ~p ~p~n", [Message, URL, Err]),
+                    {error, Message, State};
+                _:_ ->
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: This msg tells the user to look in the logs but this handler doesn't log anything.  It should log the error type, reason and the stracktrace.

However, I wonder if this catch-all is the proper thing to do.  This will catch anything: system error, process exists, user throws, etc.  Do we really want to catch all those possible errors and surpress them?  I feel like we should only deal with user throws that we expect and let everything else bubble up.

(As a side note I made bad use of throw in early versions of Yokozuna.  A lot of places that throw should instead return `{error, Reason}` tuples.  But that's outside the scope of this change).

File: src/yz_pb_search.erl [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    Value.
+
+encode_doc(Doc) ->
+    riak_pb_search_codec:encode_search_doc([{Prop, to_binary(Value)} || {Prop, Value} <- Doc]).
+
+to_binary(A) when is_atom(A) -> to_binary(atom_to_list(A));
+to_binary(B) when is_binary(B) -> B;
+to_binary(I) when is_integer(I) -> to_binary(integer_to_list(I));
+to_binary(F) when is_float(F) -> to_binary(float_to_list(F));
+to_binary(L) when is_list(L) -> unicode:characters_to_binary(L).
```
Comment (rzezeski) [inline src/yz_pb_search.erl]: This still isn't right.  The `mochijson2:decode` function will convert all JSON string to binary.  It will convert JSON arrays to a list.  So `L` will always be a list of arbitrary erlang terms.  Which means we should encode it with `term_to_binary`.  The problem is I don't think we have any easy way to make PB search smart enough to know to decode that back with `binary_to_term` on the client side.

This is important because Solr returns a multiValued store field as a JSON array.  I'm not sure how to work around this problem.

File: riak_test/yz_pb.erl [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    {Host, Port} = HP,
+    %% populate a value
+    {ok, "204", _, _} = ibrowse:send_req(URL, [{"Content-Type", "text/plain"}], put, Body),
+    %% Sleep for soft commit
+    timer:sleep(1100),
+    {ok, Pid} = riakc_pb_socket:start_link(Host, (Port-1)),
+    lager:info("search_results"),
+    {ok,{search_results,[{Bucket,Results}],Score,Found}} =
+            riakc_pb_socket:search(Pid, Bucket, Search, Params),
+    Bucket == binary_to_list(proplists:get_value(<<"_yz_rk">>, Results)),
```
Comment (rzezeski) [inline riak_test/yz_pb.erl]: This test is not actually verifying anything.  It's just doing term comparison.  I.e. '=' vs. '=='.  It makes more sense to use `?assertEqual` so you can see both expected and actual value.

File: src/yz_pb_search.erl (hunk 12 above)
Comment (coderoshi) [inline src/yz_pb_search.erl]: I added the log message, but I agree, sending an error rather than throw will be addressed in #57

File: src/yz_pb_search.erl (hunk 11 above)
Comment (coderoshi) [inline src/yz_pb_search.erl]: In my tests you'll get a `maxScore` whether you add it to the `fl` or not. The only difference with adding score to the `fl` is whether or not its pair is in the doc response.
